sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/MaxPerAttributeGenerator.java

```diff
@@ -124,4 +95,84 @@ public boolean equals(Object o) {
     public int hashCode() {
         return HashCodeBuilder.reflectionHashCode(this);
     }
+
+    /**
+     * Implementation of max attribute counts. Considers attribute usage of tasks in the cluster to
+     * determine whether the provided task can be launched against a given offer. This rule is
+     * checked at offer evaluation time in order to check that the task isn't being double-counted
+     * against itself in a config update scenario.
+     */
+    private static class MaxPerAttributeRule implements PlacementRule {
+
+        private final int maxTasksPerSelectedAttribute;
+        private final AttributeSelector attributeSelector;
+        private final Collection<TaskInfo> tasks;
+
+        private MaxPerAttributeRule(
+                int maxTasksPerSelectedAttribute,
+                AttributeSelector attributeSelector,
+                Collection<TaskInfo> tasks) {
+            this.maxTasksPerSelectedAttribute = maxTasksPerSelectedAttribute;
+            this.attributeSelector = attributeSelector;
+            this.tasks = tasks;
```

Comment:
PlacementRuleGenerator gets the current state of the system, then PlacementRule is used to evaluate offers given that state. In this case, the task list passthrough is done because we want to avoid double-counting the task that's about to be launched.

Example scenario:
- Limit per rack: 3
- Current state (Collection<TaskInfo>):
 - rack:1: tasks A, B, C
 - rack:2: tasks D, E, F
- Offer being evaluated (OfferRequirement):
 - (Re)deploy B

Without the duplicate detection, B would be unable to deploy at all because both racks are fully populated. But with the duplicate detection the old instance of B against rack:1 is filtered out because it matches the task being deployed and we'd be able to deploy B against rack:1.